### PR TITLE
Fix startup health check command behavior

### LINKED ISSUE #27724: Image value not used for health-startup-cmd
### Issue Description

For health-cmd we have:

https://github.com/containers/podman/blob/ed132b7a720e6562d54646081f80f685338f15c5/docs/source/markdown/options/health-cmd.md#L14

For health-startup-cmd there's no documentation that the configuration from the image would be used and this matches how at least 5.4.2 seems to behave https://github.com/containers/podman/blob/ed132b7a720e6562d54646081f80f685338f15c5/docs/source/markdown/options/health-startup-cmd.md

### Steps to reproduce the issue

Given an image built with buildkit:

```
$ docker inspect -f '{{json .Config.Healthcheck}}' $IMAGE
{"Test":["CMD-SHELL","pg_isready -U postgres"],"Interval":3600000000000,"StartPeriod":60000000000,"StartInterval":1000000000}
```

### Describe the results you received

With docker pg_isready is run every second:

```json
            "Health": {
                "Status": "healthy",
                "FailingStreak": 0,
                "Log": [
                    {
                        "Start": "2025-12-09T12:03:03.836191549Z",
                        "End": "2025-12-09T12:03:04.227778507Z",
                        "ExitCode": 2,
                        "Output": "/var/run/postgresql:5432 - no response\n"
                    },
                    {
                        "Start": "2025-12-09T12:03:05.229269716Z",
                        "End": "2025-12-09T12:03:05.4471183Z",
                        "ExitCode": 2,
                        "Output": "/var/run/postgresql:5432 - no response\n"
                    },
                    {
                        "Start": "2025-12-09T12:03:06.448626008Z",
                        "End": "2025-12-09T12:03:06.691341842Z",
                        "ExitCode": 0,
                        "Output": "/var/run/postgresql:5432 - accepting connections\n"
                    }
                ]
            }
```

With podman it's only run once and then scheduled 

```
2025-12-09 11:14:41.340943129 +0000 UTC container health_status feeb1bb73a5b00ea57ace734c4d4def9a4cb76631f1a2b2877d0655fc47a0005 (image=IMAGE, name=postgres, health_status=starting, health_failing_streak=0, health_log=, PODMAN_SYSTEMD_UNIT=postgres.service)
```

### Describe the results you expected

Podman behavior should match Docker.

### podman info output

```yaml
host:
  arch: amd64
  buildahVersion: 1.39.3
  cgroupControllers:
  - cpuset
  - cpu
  - io
  - memory
  - hugetlb
  - pids
  - rdma
  - misc
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: conmon_2.1.12-4_amd64
    path: /usr/bin/conmon
    version: 'conmon version 2.1.12, commit: unknown'
  cpuUtilization:
    idlePercent: 98.53
    systemPercent: 0.5
    userPercent: 0.96
  cpus: 2
  databaseBackend: sqlite
  distribution:
    codename: trixie
    distribution: debian
    version: "13"
  eventLogger: journald
  freeLocks: 2046
  hostname: postgres
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 6.12.41+deb13-cloud-amd64
  linkmode: dynamic
  logDriver: journald
  memFree: 1439248384
  memTotal: 4012314624
  networkBackend: netavark
  networkBackendInfo:
    backend: netavark
    dns:
      package: Unknown
    package: netavark_1.14.0-2_amd64
    path: /usr/lib/podman/netavark
    version: netavark 1.14.0
  ociRuntime:
    name: crun
    package: crun_1.21-1_amd64
    path: /usr/bin/crun
    version: |-
      crun version 1.21
      commit: 10269840aa07fb7e6b7e1acff6198692d8ff5c88
      rundir: /run/user/0/crun
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +WASM:wasmedge +YAJL
  os: linux
  pasta:
    executable: ""
    package: ""
    version: ""
  remoteSocket:
    exists: true
    path: /run/podman/podman.sock
  rootlessNetworkCmd: pasta
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: false
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: false
  slirp4netns:
    executable: ""
    package: ""
    version: ""
  swapFree: 0
  swapTotal: 0
  uptime: 0h 44m 50.00s
  variant: ""
plugins:
  authorization: null
  log:
  - k8s-file
  - none
  - passthrough
  - journald
  network:
  - bridge
  - macvlan
  - ipvlan
  volume:
  - local
registries: {}
store:
  configFile: /usr/share/containers/storage.conf
  containerStore:
    number: 1
    paused: 0
    running: 1
    stopped: 0
  graphDriverName: overlay
  graphOptions: {}
  graphRoot: /var/lib/containers/storage
  graphRootAllocated: 39990112256
  graphRootUsed: 1925476352
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Supports shifting: "true"
    Supports volatile: "true"
    Using metacopy: "false"
  imageCopyTmpDir: /var/tmp
  imageStore:
    number: 1
  runRoot: /run/containers/storage
  transientStore: false
  volumePath: /var/lib/containers/storage/volumes
version:
  APIVersion: 5.4.2
  BuildOrigin: Debian
  Built: 1753478586
  BuiltTime: Fri Jul 25 21:23:06 2025
  GitCommit: ""
  GoVersion: go1.24.4
  Os: linux
  OsArch: linux/amd64
  Version: 5.4.2
```

### Podman in a container

No

### Privileged Or Rootless

None

### Upstream Latest Release

Yes

### Additional environment details

Additional environment details

### Additional information

If later release has fixed the behaviour, at minimum the documentation in main need updating.

### PR DESCRIPTION
**Pull request by arsenalzp**
The issue is that health check command is ignored when options `--health-start-period`, `--health-startup-interval` are set, but `--health-startup-cmd` is not present.
In contrast to `Docker` which uses health check command for the startup health check.

#### Checklist

Ensure you have completed the following checklist for your pull request to be reviewed:

- [x] Certify you wrote the patch or otherwise have the right to pass it on as an open-source patch by signing all
commits. (`git commit -s`). (If needed, use `git commit -s --amend`).  The author email must match
the sign-off email address. See [CONTRIBUTING.md](https://github.com/containers/podman/blob/main/CONTRIBUTING.md#sign-your-prs)
for more information.
- [x] Referenced issues using `Fixes: #00000` in commit message (if applicable)
- [x] [Tests](https://github.com/containers/podman/tree/main/test#readme) have been added/updated (or no tests are needed)
- [x] [Documentation](https://github.com/containers/podman/blob/main/docs/README.md) has been updated (or no documentation changes are needed)
- [x] All commits pass `make validatepr` (format/lint checks)
- [x] [Release note](https://github.com/kubernetes/community/blob/master/contributors/guide/release-notes.md) entered in the section below (or `None` if no user-facing changes)

#### Does this PR introduce a user-facing change?

Yes
```release-note
Fix `--health-startup-cmd` option behavior when its the value is not set
```
Fixes: #27724.

### CODE REVIEW

File: pkg/specgenutil/specgen.go [hunk 1]
```diff
@@ -395,6 +395,31 @@ func FillOutSpecGen(s *specgen.SpecGenerator, c *entities.ContainerCreateOptions
 		s.StartupHealthConfig.Successes = int(c.StartupHCSuccesses)
 	}
 
+	// It is necessary to parse time values, as user can
+	// set either "0" or "0s" arguments
+	StartupHCInterval, _ := time.ParseDuration(c.StartupHCInterval)
```
Comment (baude) [inline pkg/specgenutil/specgen.go]: should these errors not be handled?

Comment (arsenalzp) [reply pkg/specgenutil/specgen.go]: It seems values are checked in earlier stages.
However, let me double check it.

Comment (arsenalzp) [reply pkg/specgenutil/specgen.go]: @baude the idea was to get value from the string parameter in case a user set a time by using values: "0", "0s".
Otherwise it is necessary to check each case:
```
if value == "0"{}
if value == "0s" {}
```

**Comment by TomSweeneyRedHat** [comment]
I've restarted what appears to be flakey tests.  The changes LGTM, but I definitely want a head nod from @mheon

**Comment by arsenalzp** [comment]
Hello @mheon,
Could you please be so kind to review this PR?

**Comment by mheon** [comment]
I think you need to add an alias for `health-start-interval` to `health-startup-interval` for Docker compatibility, and I don't see handling for the new field in image health check configs?

File: pkg/specgenutil/specgen.go (hunk 1 above)
Comment (mheon) [inline pkg/specgenutil/specgen.go]: Why are errors ignored here?

Comment (arsenalzp) [reply pkg/specgenutil/specgen.go]: Hello,
the idea was to get value from the string parameter in case a user set a time by using values: "0", "0s".
Otherwise it is necessary to check each case:

```
if value == "0"{}
if value == "0s" {}
```

**Comment by arsenalzp** [comment]
> I think you need to add an alias for `health-start-interval` to `health-startup-interval` for Docker compatibility, and I don't see handling for the new field in image health check configs?

Hello,
I agree with creating alias for parameters.
Regarding the image health check config -  do you have any suggestion how it would be implemented in the good way? I will do that.
Thank you!

**Comment by TheRealBecks** [comment]
@mheon Any idea how to proceed here?

**Comment by Honny1** [comment]
@arsenalzp I think Matt meant that the `--start-period` flag is set from the image (`HEALTHCHECK` command). Here is the code that performs the override of image hc: https://github.com/containers/podman/blob/f7aefa46bbe5f0d8b1cfb0ff7b9fcd5fee2a9545/pkg/specgen/generate/container.go#L65

**Comment by Honny1** [comment]
@arsenalzp Any updates on this?

**Comment by arsenalzp** [comment]
> @arsenalzp Any updates on this?

Hello,
Yes, I remember about this PR.
I will take a look at it soon!

**Comment by arsenalzp** [comment]
Hello,
It this PR still valid? Can I proceed work on it?

**Comment by Honny1** [comment]
Yes

**Comment by arsenalzp** [comment]
It was a quite time ago so, as far as I understand, we need to implement overriding of an image health check value by one from CLI.

File: pkg/specgenutil/specgen.go [hunk 2]
```diff
@@ -394,6 +394,31 @@ func FillOutSpecGen(s *specgen.SpecGenerator, c *entities.ContainerCreateOptions
...
+	HealthInterval, _ := time.ParseDuration(c.HealthInterval)
+	HealthStartPeriod, _ := time.ParseDuration(c.HealthStartPeriod)
+
+	// This part of code has been created on response to the issue #27724:
+	if c.StartupHCCmd == "" &&
+		StartupHCInterval != HealthInterval &&
+		HealthStartPeriod != time.Duration(0) &&
```
Comment (Honny1) [inline pkg/specgenutil/specgen.go]: I would use a more robust approach and propagate `cmd.Flags().Changed("health-startup-interval")` Or make a pointer to that value.

File: libpod/define/healthchecks.go [hunk 3]
```diff
@@ -50,6 +50,9 @@ const (
 	// HealthCheckStartup means the healthcheck was unhealthy, but is still
 	// either within the startup HC or the startup period of the healthcheck
 	HealthCheckStartup HealthCheckStatus = iota
+	// HealthCheckStartup means the healthcheck was unhealthy and it is out
+	// of start period
+	HealthCheckStartupFailure HealthCheckStatus = iota
```
Comment (Honny1) [inline libpod/define/healthchecks.go:55]: Is it handled in `String()`?

File: libpod/healthcheck.go [hunk 4]
```diff
@@ -155,9 +155,11 @@ func (c *Container) runHealthCheck(ctx context.Context, isStartup bool) (define.
 
 	// Handle startup HC
 	if isStartup {
-		inStartPeriod = true
```
Comment (Honny1) [inline libpod/healthcheck.go:158]: This will change the behavior of subsequent function calls.

**Review by Honny1 — COMMENTED** [review]
I did review. My memory is rusty but based on converstion is needed:
- adding `--health-start-interval` as an alias for `--health-startup-interval` for Docker compatibility.
- the `--start-period` and other flags can be seted from the image (HEALTHCHECK command). Here is the code that performs the override of image hc: https://github.com/containers/podman/blob/f7aefa46bbe5f0d8b1cfb0ff7b9fcd5fee2a9545/pkg/specgen/generate/container.go#L65


Note for me: Next review check whole hc logic.

File: libpod/healthcheck.go (hunk 4 above)
Comment (arsenalzp) [reply libpod/healthcheck.go:158]: No, it will not. Otherwise container would be infinitely in the start period. Anyway, let me double check the algorithm.

Comment (arsenalzp) [reply libpod/healthcheck.go:158]: @Honny1 I've just checked, here is where startup period is set:
https://github.com/arsenalzp/podman/blob/f791617e0382100886bb6cde8d2afc8e23404fab/libpod/healthcheck.go#L74-L78

Then the startup period is always true as it is overrode here:
https://github.com/arsenalzp/podman/blob/86bf912fc5c719f368a07f080e83c7f046a994fe/libpod/healthcheck.go#L157-L158

That's why you see a container status is "starting", despite of startup period ended long time ago.

I removed the one line of code, therefore if the `inStartPeriod` variable wasn't set to **true** here:
```
if timeStart.Before(startPeriodTime) {
			// we are still in the start period, flip the inStartPeriod bool
			inStartPeriod = true
			logrus.Debugf("healthcheck for %s being run in start-period", c.ID())
}
```
then `inStartPeriod` was set to the default value **false** means we are ran out from startup period and a container has to fall in `unhealthy` state (if a command failed, of course).

File: libpod/define/healthchecks.go (hunk 3 above)
Comment (arsenalzp) [reply libpod/define/healthchecks.go:55]: Yes, the value `HealthCheckStartupFailure` is handled as default case, means `unhealthy`

File: libpod/healthcheck.go (hunk 4 above)
Comment (Honny1) [reply libpod/healthcheck.go:158]: Oh, I see. Thanks

File: pkg/specgenutil/specgen.go (hunk 2 above)
Comment (arsenalzp) [reply pkg/specgenutil/specgen.go]: It would be good to have, but I'm afraid it will make drastically changes in the code as all those values go through multiple function calls until come to `FillOutSpecGen` function, therefore one is called 12 times from other places.

Comment (Honny1) [reply pkg/specgenutil/specgen.go]: Sure, good point

**Comment by Honny1** [comment]
Well Docker's `--health-start-interval` is not a separate startup healthcheck.  It simply runs the same health command at a different (shorter) interval during the start period. Podman's startup healthcheck (`--health-startup-cmd`) is a fundamentally different concept: an entirely separate command that must succeed before the main HC begins. 

This PR mixes StartupHC and HC by creating a `StartupHealthConfig` using the regular `--health-cmd` when `--health-startup-interval` is set. This means the container will run the Podman StartupHC machinery (which has its own success/failure counters, transitions, etc.) rather than simply running the health command at a shorter interval during the start period.  

I would not mix `--health-start-period` and `--health-startup-interval`.

**Comment by arsenalzp** [comment]
Hello, no this PR doesn't. 
As we discussed in the linked issue, the main purpose of this PR was to use health check command for start-up health check purposes in case when start-up command wasn't set. So, if you set start-up command then it is used, otherwise common health check command is used.

The second issue the PR fixes - it is startup interval, which doesn't work. You can check corresponding issue link, there are plenty snippets which show how does Docker work and Podman doesn't.

Let's summarize:
- health check command is used when star-up health check command wasn't set - done;
- start-up health check interval in the start-up health check period is now working as expected;

Here is the [observation](https://github.com/podman-container-tools/podman/issues/27724#issuecomment-3690325397)

**Comment by arsenalzp** [comment]
@Honny1 let's take a look on the fix was done by my PR:

```
$ bin/podman run -d --name test --health-cmd 'stat /tmp/true' --health-start-period 12s --health-startup-interval 4s --stop-signal SIGTERM quay.io/libpod/testimage:20241011 sh -c 'sleep 8 && touch /tmp/true && sleep infinity'
```
and now check the health:

```
$ bin/podman inspect -f '{{json .State.Health}}' test | jq
{
  "Status": "starting",
  "FailingStreak": 0,
  "Log": [
    {
      "Start": "2026-08-25T22:47:38.734941826+03:00",
      "End": "2026-08-25T22:47:38.796719315+03:00",
      "ExitCode": 1,
      "Output": "stat: can't stat '/tmp/true': No such file or directory\n"
    },
    {
      "Start": "2026-08-25T22:47:43.378029581+03:00",
      "End": "2026-08-25T22:47:43.426019666+03:00",
      "ExitCode": 1,
      "Output": "stat: can't stat '/tmp/true': No such file or directory\n"
    }
  ]
}
```

and bit later:

```
$ bin/podman inspect -f '{{json .State.Health}}' test | jq
{
  "Status": "healthy",
  "FailingStreak": 0,
  "Log": [
    {
      "Start": "2026-08-25T22:47:38.734941826+03:00",
      "End": "2026-08-25T22:47:38.796719315+03:00",
      "ExitCode": 1,
      "Output": "stat: can't stat '/tmp/true': No such file or directory\n"
    },
    {
      "Start": "2026-08-25T22:47:43.378029581+03:00",
      "End": "2026-08-25T22:47:43.426019666+03:00",
      "ExitCode": 1,
      "Output": "stat: can't stat '/tmp/true': No such file or directory\n"
    },
    {
      "Start": "2026-08-25T22:47:48.791204551+03:00",
      "End": "2026-08-25T22:47:48.812022621+03:00",
      "ExitCode": 0,
      "Output": "  File: /tmp/true\n  Size: 0         \tBlocks: 0          IO Block: 4096   regular empty file\nDevice: 1000ach/1048748d\tInode: 2097359     Links: 1\nAccess: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)\nAccess: 2026-08-25 19:47:46.440184471 +0000\nModify: 2026-08-25 19:47:46.440184471 +0000\nChange: 2026-08-25 19:47:46.440184471 +0000\n"
    }
  ]
}
```

So, the container's start-up health is checked every 4 sec, then the container was ready in 8 sec (the file was created) and the start-up health check has been transited to common health check algorithm which runs every 30 sec:

```
$ bin/podman inspect -f '{{json .State.Health}}' test | jq
{
  "Status": "healthy",
  "FailingStreak": 0,
  "Log": [
    {
      "Start": "2026-08-25T22:51:25.65021046+03:00",
      "End": "2026-08-25T22:51:25.675402972+03:00",
      "ExitCode": 0,
      "Output": "  File: /tmp/true\n  Size: 0         \tBlocks: 0          IO Block: 4096   regular empty file\nDevice: 1000ach/1048748d\tInode: 2097359     Links: 1\nAccess: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)\nAccess: 2026-08-25 19:47:46.440184471 +0000\nModify: 2026-08-25 19:47:46.440184471 +0000\nChange: 2026-08-25 19:47:46.440184471 +0000\n"
    },
    {
      "Start": "2026-08-25T22:51:56.534258393+03:00",
      "End": "2026-08-25T22:51:56.584108198+03:00",
      "ExitCode": 0,
      "Output": "  File: /tmp/true\n  Size: 0         \tBlocks: 0          IO Block: 4096   regular empty file\nDevice: 1000ach/1048748d\tInode: 2097359     Links: 1\nAccess: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)\nAccess: 2026-08-25 19:47:46.440184471 +0000\nModify: 2026-08-25 19:47:46.440184471 +0000\nChange: 2026-08-25 19:47:46.440184471 +0000\n"
    },
    {
      "Start": "2026-08-25T22:52:27.685199688+03:00",
      "End": "2026-08-25T22:52:27.748304981+03:00",
      "ExitCode": 0,
      "Output": "  File: /tmp/true\n  Size: 0         \tBlocks: 0          IO Block: 4096   regular empty file\nDevice: 1000ach/1048748d\tInode: 2097359     Links: 1\nAccess: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)\nAccess: 2026-08-25 19:47:46.440184471 +0000\nModify: 2026-08-25 19:47:46.440184471 +0000\nChange: 2026-08-25 19:47:46.440184471 +0000\n"
    },
    {
      "Start": "2026-08-25T22:52:58.647056082+03:00",
      "End": "2026-08-25T22:52:58.694411579+03:00",
      "ExitCode": 0,
      "Output": "  File: /tmp/true\n  Size: 0         \tBlocks: 0          IO Block: 4096   regular empty file\nDevice: 1000ach/1048748d\tInode: 2097359     Links: 1\nAccess: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)\nAccess: 2026-08-25 19:47:46.440184471 +0000\nModify: 2026-08-25 19:47:46.440184471 +0000\nChange: 2026-08-25 19:47:46.440184471 +0000\n"
    },
    {
      "Start": "2026-08-25T22:53:29.720210717+03:00",
      "End": "2026-08-25T22:53:29.738156295+03:00",
      "ExitCode": 0,
      "Output": "  File: /tmp/true\n  Size: 0         \tBlocks: 0          IO Block: 4096   regular empty file\nDevice: 1000ach/1048748d\tInode: 2097359     Links: 1\nAccess: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)\nAccess: 2026-08-25 19:47:46.440184471 +0000\nModify: 2026-08-25 19:47:46.440184471 +0000\nChange: 2026-08-25 19:47:46.440184471 +0000\n"
    }
  ]
}
```

Now, let's set the start-up health command:
```
bin/podman run -d --name test --health-startup-cmd 'stat /tmp/true' --health-start-period 12s --health-startup-interval 4s --health-cmd 'echo Im healthy' --stop-signal SIGTERM quay.io/libpod/testimage:20241011 sh -c 'sleep 9 && touch /tmp/true && sleep infinity'
```

and then check the health:

```
$ bin/podman inspect -f '{{json .State.Health}}' test | jq
{
  "Status": "healthy",
  "FailingStreak": 0,
  "Log": [
    {
      "Start": "2026-08-25T23:06:22.813961771+03:00",
      "End": "2026-08-25T23:06:22.857790506+03:00",
      "ExitCode": 1,
      "Output": "stat: can't stat '/tmp/true': No such file or directory\n"
    },
    {
      "Start": "2026-08-25T23:06:27.383670962+03:00",
      "End": "2026-08-25T23:06:27.40040345+03:00",
      "ExitCode": 1,
      "Output": "stat: can't stat '/tmp/true': No such file or directory\n"
    },
    {
      "Start": "2026-08-25T23:06:32.659081492+03:00",
      "End": "2026-08-25T23:06:32.712440158+03:00",
      "ExitCode": 0,
      "Output": "Im healthy\n"
    },
    {
      "Start": "2026-08-25T23:07:03.388737924+03:00",
      "End": "2026-08-25T23:07:03.44015703+03:00",
      "ExitCode": 0,
      "Output": "Im healthy\n"
    }
  ]
}
```

That's exactly what was expected!

-----------

And now, let's take a look how does it work with the current algorithm (`podman Version:  6.2.0-dev`):

```
$ bin/podman run -d --name test --health-cmd 'stat /tmp/true' --health-start-period 12s --health-startup-interval 4s --stop-signal SIGTERM quay.io/libpod/testimage:20241011 sh -c 'sleep 8 && touch /tmp/true && sleep infinity'
```

let's check the health:

```
$ bin/podman inspect -f '{{json .State.Health}}' test | jq
{
  "Status": "healthy",
  "FailingStreak": 0,
  "Log": [
    {
      "Start": "2026-08-25T23:14:31.452404124+03:00",
      "End": "2026-08-25T23:14:31.494189713+03:00",
      "ExitCode": 1,
      "Output": "stat: can't stat '/tmp/true': No such file or directory\n"
    },
    {
      "Start": "2026-08-25T23:15:02.341189475+03:00",
      "End": "2026-08-25T23:15:02.392869195+03:00",
      "ExitCode": 0,
      "Output": "  File: /tmp/true\n  Size: 0         \tBlocks: 0          IO Block: 4096   regular empty file\nDevice: 1000ach/1048748d\tInode: 2097359     Links: 1\nAccess: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)\nAccess: 2026-08-25 20:14:39.322574021 +0000\nModify: 2026-08-25 20:14:39.322574021 +0000\nChange: 2026-08-25 20:14:39.322574021 +0000\n"
    }
  ]
}
```

As you may see the command runs every 30 sec, despite we set the start-up health interval to 4 sec.

Another discrepancy: 

```
bin/podman run -d --name test --health-cmd /bin/false --health-start-period 12s --health-startup-interval 4s --stop-signal SIGTERM quay.io/libpod/testimage:20241011 sh -c 'sleep infinity'
```

let's check the health in 10 sec:

```
$ bin/podman inspect -f '{{json .State.Health}}' test | jq
{
  "Status": "starting",
  "FailingStreak": 2,
  "Log": [
    {
      "Start": "2026-08-25T23:21:47.457388804+03:00",
      "End": "2026-08-25T23:21:47.605835246+03:00",
      "ExitCode": 1,
      "Output": ""
    },
    {
      "Start": "2026-08-25T23:22:18.2958529+03:00",
      "End": "2026-08-25T23:22:18.387919946+03:00",
      "ExitCode": 1,
      "Output": ""
    },
    {
      "Start": "2026-08-25T23:22:48.907862571+03:00",
      "End": "2026-08-25T23:22:48.985084484+03:00",
      "ExitCode": 1,
      "Output": ""
    }
  ]
}
```
Hmm, the container is still starting, despite of the start-up period was set to 8 sec. The start-up health interval was to 4 sec, however the health check runs every 30 sec (default).

File: libpod/define/healthchecks.go (hunk 3 above)
Comment (Honny1) [reply libpod/define/healthchecks.go:55]: Is there a difference between this and `HealthCheckFailure`?

File: test/e2e/healthcheck_run_test.go [hunk 5]
```diff
@@ -425,4 +425,14 @@ HEALTHCHECK CMD ls -l / 2>&1`, ALPINE)
 		hc.WaitWithTimeout(10)
 		Expect(hc).Should(ExitWithError(125, "Error: healthcheck command exceeded timeout of 3s"))
 	})
+
+	It("podman healthcheck startup command is not set with startup period", func() {
```
Comment (Honny1) [inline test/e2e/healthcheck_run_test.go:429]: Can you please add a test for this command:
```
podman run -d --name test-regression \
  --health-cmd "stat /tmp/ready" \
  --health-startup-cmd "stat /tmp/ready" \
  --health-startup-interval 3s \
  --health-start-period 5s \
  --health-retries 3 \
  quay.io/libpod/testimage:20241011 \
  sh -c 'sleep 30 && touch /tmp/ready && sleep infinity'
```

Also, tests should check that multiple health check log entries appear within the expected startup interval window.

File: libpod/define/healthchecks.go [hunk 6]
```diff
@@ -74,6 +77,8 @@ const (
 	DefaultHealthCheckRetries uint = 3
 	// DefaultHealthCheckStartPeriod default value
 	DefaultHealthCheckStartPeriod = "0s"
+	// DefaultHealthCheckStartInterval default value
+	DefaultHealthCheckStartInterval = "0s"
```
Comment (Honny1) [inline libpod/define/healthchecks.go:81]: Docker has `5s`.

File: pkg/specgenutil/specgen.go [hunk 7]
```diff
@@ -394,6 +394,37 @@ func FillOutSpecGen(s *specgen.SpecGenerator, c *entities.ContainerCreateOptions
 		s.StartupHealthConfig.Successes = int(c.StartupHCSuccesses)
 	}
 
+	// It is necessary to parse time values, as user can
+	// set either "0" or "0s" arguments
+	StartupHCInterval, _ := time.ParseDuration(c.StartupHCInterval)
```
Comment (Honny1) [inline pkg/specgenutil/specgen.go]: I would name variables according to Go conventions. 
```suggestion
	startupHCInterval, _ := time.ParseDuration(c.StartupHCInterval)
```

**Review by Honny1 — CHANGES_REQUESTED** [review]
Please squash commits and properly link the issue number in the commit (`Fixes: #number`). Also, don't forget to update the documentation.

File: libpod/define/healthchecks.go (hunk 3 above)
Comment (arsenalzp) [reply libpod/define/healthchecks.go:55]: No, there is no difference at this moment, however it would be better to distinguish which kind of health probe failed.
1) the health check result is used in `container.updateHealthCheckLog()` method so it would be easier to distinguish the kind of health probe failure, and either update info with appropriate value or return distinguished RC.
2) the code is more clear, so instead of short/long text comment we use the const:
```
if isStartup {
		if hcErr != nil || exitCode != 0 {
			hcResult = define.HealthCheckStartup
			if !inStartPeriod {
				hcResult = define.HealthCheckStartupFailure
			}
			if err := c.incrementStartupHCFailureCounter(ctx); err != nil {
				return define.HealthCheckInternalError, "", err
			}
		}
```

File: libpod/define/healthchecks.go (hunk 6 above)
Comment (arsenalzp) [reply libpod/define/healthchecks.go:81]: I've double checked the docker manual, the latest version states the default value is "0s":

```
--health-start-interval		API 1.44+ Time between running the check during the start period (ms|s|m|h) (default 0s)
```

File: pkg/specgenutil/specgen.go (hunk 7 above)
Comment (arsenalzp) [reply pkg/specgenutil/specgen.go]: Yes, that was my typos, fixed the var `healthInterval` as well

File: test/e2e/healthcheck_run_test.go (hunk 5 above)
Comment (arsenalzp) [reply test/e2e/healthcheck_run_test.go:429]: Yes, I can add another test.
However, as health checks are ran by systemd timer unit, the accuracy might differ on different system. The value of `AccuracySec` is **1s**, therefore the inaccuracy of health check interval is  **1 sec** + time on health check command execution in an environment.
Decreasing of  `AccuracySec` value leads to more accuracy, the timer unit runs within health check startup interval value in 7 of 10 cases on my workstation.

File: libpod/define/healthchecks.go (hunk 6 above)
Comment (Honny1) [reply libpod/define/healthchecks.go:81]: Ah, I see. I was refering to this: https://docs.docker.com/reference/dockerfile/#healthcheck:~:text=%2D%2Dstart%2Dinterval%3DDURATION%20(default%3A%205s)

File: test/e2e/healthcheck_run_test.go (hunk 5 above)
Comment (Honny1) [reply test/e2e/healthcheck_run_test.go:429]: I would adjust values to account for this type of inaccuracy. (Also, I chose this test to check an edge case.)

**Review by Honny1 — COMMENTED** [review]
Can you please adjust the commit msg to this format:

```
<subject>

Fixes: #number
Signed-off-by: ...
```

File: test/e2e/healthcheck_run_test.go (hunk 5 above)
Comment (arsenalzp) [reply test/e2e/healthcheck_run_test.go:429]: That's what I did with `Eventually` calls:
1) during the start period the container is in `starting` condition
2) once the start period ends in 5 sec, the container gets into `unhealthy` condition
3) finally, the container is in `healthy` condition in 30 (25+5) sec

**Comment by arsenalzp** [comment]
@Honny1 it seems, all remarks have been taken into account and removed. Could this PR be merged?

File: docs/source/markdown/options/health-startup-cmd.md [hunk 8]
```diff
@@ -13,4 +13,5 @@ healthcheck. When the startup command succeeds, the regular healthcheck begins a
 if the command fails a set number of times, the container is restarted. A startup healthcheck can be used to ensure that
 containers with an extended startup period are not marked as unhealthy until they are fully started. Startup healthchecks can only be
 used when a regular healthcheck (from the container's image or the
-<< '`HealthCmd=`' if is_quadlet else '`--health-cmd`' >> option) is also set.
```
Comment (mheon) [inline docs/source/markdown/options/health-startup-cmd.md:16]: I think you need the special << >> behavior here so things work out properly when the option is used in Quadlet manpages

Comment (mheon) [reply docs/source/markdown/options/health-startup-cmd.md:16]: Same before for the `--health-cmd` and `--health-startup-cmd` bits

File: docs/source/markdown/options/health-startup-interval.md [hunk 9]
```diff
@@ -8,4 +8,4 @@
 #### **--health-startup-interval**=*interval*
 << endif >>
 
-Set an interval for the startup healthcheck. An _interval_ of **disable** results in no automatic timer setup. The default is **30s**.
+Set an interval for the startup healthcheck. An _interval_ of **disable** results in no automatic timer setup. The default is **0s**.
```
Comment (mheon) [inline docs/source/markdown/options/health-startup-interval.md:11]: How is 0s different from disable? This is a bit confusing as written

**Review by mheon — APPROVED** [review]
Bit of work needed on the manpages. Code and tests LGTM

**Review by Honny1 — APPROVED** [review]
LGTM, onece manpages are fixed.

**Review by Honny1 — APPROVED** [review]
LGTM

**Comment by Honny1** [comment]
@podman-container-tools/podman-maintainers PTAL and merge

File: test/e2e/healthcheck_run_test.go [hunk 10]
```diff
@@ -425,4 +425,38 @@ HEALTHCHECK CMD ls -l / 2>&1`, ALPINE)
...
+			hc.WaitWithTimeout(2)
+			return hc.OutputToString()
+		}, 6*time.Second, 5*time.Second).To(BeEquivalentTo("unhealthy"))
+
+		Eventually(func() int {
+			hc := podmanTest.Podman([]string{"healthcheck", "run", ctrName})
+			hc.WaitWithTimeout(2)
+			return hc.ExitCode()
+		}, 26*time.Second, 25*time.Second).Should(Equal(0))
+	})
```
Comment (Luap99) [inline test/e2e/healthcheck_run_test.go:461]: This test takes 30 seconds which is unreasonable and taking to much CI time.
The polling interval of 5 and 25 seconds are just way to long with a sleep of 30 seconds. This needs to be cut down ideally to less than 5 seconds.

**Review by Luap99 — CHANGES_REQUESTED** [review]

**Review by Luap99 — APPROVED** [review]
LGTM, thanks
